typing/oprint.ml

```diff
@@ -459,38 +459,65 @@ let out_sig_item = ref (fun _ -> failwith "Oprint.out_sig_item")
-  | Omty_functor (name, Some mty_arg, mty_res) -> begin
-      match name, funct with
-      | "_", true ->
-          fprintf ppf "->@ %a ->@ %a"
-            print_simple_out_module_type mty_arg
-            (print_out_functor false) mty_res
-      | "_", false ->
+let rec print_out_module_type ppf mty =
+  print_out_functor ppf mty
+and print_out_functor ppf = function
+  | Omty_functor _ as t ->
+     (* For anonymous functor arguments, the logic to choose between
+        the long-form
+          functor (_ : S) -> ...
+        and the short-form
+          S -> ...
+        is as follows: if we are already printing long-form functor arguments,
+        we use the long form unless all remaining functor arguments can use
+        the short form. (Otherwise use the short form.)
+
+        For example,
+          functor (X : S1) (_ : S2) (Y : S3) (_ : S4) (_ : S5) -> sig end
+        will get printed as
+          functor (X : S1) (_ : S2) (Y : S3) -> S4 -> S5 -> sig end
+
+        but
+          functor (_ : S1) (_ : S2) (Y : S3) (_ : S4) (_ : S5) -> sig end
+        gets printed as
+          S1 -> S2 -> functor (Y : S3) -> S4 -> S5 -> sig end
+      *)
+     let rec print_functor ppf = function
+       | Omty_functor ("_", Some mty_arg, mty_res) ->
           fprintf ppf "%a ->@ %a"
             print_simple_out_module_type mty_arg
-            (print_out_functor false) mty_res
-      | name, true ->
-          fprintf ppf "(%s : %a)@ %a" name
-            print_out_module_type mty_arg (print_out_functor true) mty_res
-      | name, false ->
-            fprintf ppf "functor@ (%s : %a)@ %a" name
-              print_out_module_type mty_arg (print_out_functor true) mty_res
-    end
-  | m ->
-      if funct then fprintf ppf "->@ %a" print_out_module_type m
-      else print_out_module_type ppf m
-and print_out_module_type ppf = function
-  | Omty_functor _ as t -> fprintf ppf "@[<2>%a@]" (print_out_functor false) t
+            print_functor mty_res
+       | Omty_functor _ as non_anonymous_functor ->
+          let rec split_args acc = function
+            | Omty_functor (name, mty_arg, mty_res) as func_mty
+                 when not (only_anonymous_args func_mty) ->
+               split_args ((name, mty_arg) :: acc) mty_res
+            | rest -> (List.rev acc, rest)
+          and only_anonymous_args = function
+            | Omty_functor (name, _, mty_res) ->
+               name = "_" && only_anonymous_args mty_res
+            | _ -> true
+          in
+          let print_arg ppf = function
+            | (_, None) ->
+               fprintf ppf "()"
+            | (name, Some mty) ->
+               fprintf ppf "(%s : %a)"
+                 name
+                 print_out_module_type mty
+          in
+          let (args, rest) = split_args [] non_anonymous_functor in
+          fprintf ppf "functor@ @[%a@]@ ->@ %a"
+            (pp_print_list ~pp_sep:pp_print_space print_arg) args
+            print_functor rest
```

Comment:
Using `print_functor` here for the body means that, if there are further functor arrows remaining (consider `functor (X : A) -> B -> C -> sig end` for example), we will remain in the inner "print a functor" logic without opening another box.

Reply:
Compared with the current formatting, I am not sure about the location of the box in `functor@ @[%a@]...`: with this box, the formatting engine will try to fit all long form arguments on the same line, adding a line break after `functor` if it helps breaking the list of argument a little bit later. Contrarily, the previous formatting hints put the break hint between `functor` and the first argument at the same level than the breaks between arguments. This is the reason behind most of the formatting changes in the illegal_permutation test for instance. Can you comment on this change?